hw/dv/dpi/dmidpi/dmidpi.c

```diff
@@ -0,0 +1,410 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "dmidpi.h"
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+/**
+ * Setup the correct shift register data
+ *
+ * @param ctx dmidpi context object
+ */
+static void set_dr_data(struct dmidpi_ctx *ctx) {
+  switch (ctx->jtag.ir_captured) {
+    case Bypass0:
+    case Bypass1:
+      ctx->jtag.dr_shift_reg = 0;
+      ctx->jtag.dr_length = 1;
+      return;
+    case IdCode:
+      ctx->jtag.dr_shift_reg = 0x04F5484D;
+      ctx->jtag.dr_length = 32;
+      return;
+    case DTMCSR:
+      ctx->jtag.dr_shift_reg = 0x00000071;
+      ctx->jtag.dr_length = 32;
+      return;
+    case DMIAccess:
+      ctx->jtag.dr_shift_reg = ctx->jtag.dr_captured;
+      ctx->jtag.dr_length = 32 + 7 + 2;
+      return;
+    default:
+      return;
+  }
+}
+
+/**
+ * Drive a new DMI transaction to the DPI interface
+ *
+ * @param ctx dmidpi context object
+ */
+static void issue_dmi_req(struct dmidpi_ctx *ctx) {
```

Comment:
Can you align the naming of this function with `process_dmi_signals()` to make it clear that they belong together. And possibly move them closer to each other in this source file.

Reply:
I haven't managed to resolve this. The `process_dmi_signals()` function is called every cycle to update inputs from the design. The `issue_dmi_req()` function is called at specific times from the JTAG state machine below. But, it does update outputs to the design.